### PR TITLE
BrokerConsumerStats - refactored the server side logic to get stats.

### PR DESCRIPTION
**Pull request by jai1**
### Motivation
Don't need to pass topic name or subscription name to get broker side consumer stats.

### Modifications
a. Changed proto file 
b. refactored the server side logic to get stats.

### Result

No visible change on client side.

### CODE REVIEW

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/ServerCnx.java [hunk 1]
```diff
@@ -232,61 +232,30 @@ protected void handleConsumerStats(CommandConsumerStats commandConsumerStats) {
...
-        final String subscriptionName = commandConsumerStats.getSubscriptionName();
         final long consumerId = commandConsumerStats.getConsumerId();
-        
+        CompletableFuture<Consumer> consumerFuture = consumers.get(consumerId);
+        Consumer consumer = consumerFuture.getNow(null);
+        ByteBuf msg = null;
+
         if (log.isDebugEnabled()) {
-            log.debug("CommandConsumerStats[requestId = {}, topicName = {}, subscriptionName = {}, consumerId = {}]", requestId, topicName, subscriptionName, consumerId);
+            log.debug("CommandConsumerStats[requestId = {}, consumerId = {}]",
```
Comment (saandrews) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/ServerCnx.java]: We can log topic, sub name here.

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/ServerCnx.java [hunk 2]
```diff
@@ -301,9 +270,19 @@ protected void handleConsumerStats(CommandConsumerStats commandConsumerStats) {
...
         
-        commandConsumerStatsResponseBuilder.setType(subscription.getTypeString());
-        commandConsumerStatsResponseBuilder.setMsgRateExpired(subscription.getExpiredMessageRate());
+        Subscription subscription = consumer.getSubscription();
         commandConsumerStatsResponseBuilder.setMsgBacklog(subscription.getNumberOfEntriesInBacklog());
+        
+        String subsTypeString = "Null";
+        double msgExpireRate = -1; 
+        if (subscription instanceof PersistentSubscription) {
+            msgExpireRate = ((PersistentSubscription)subscription).getExpiredMessageRate();
```
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/ServerCnx.java]: instead casting to `PersistentSubscription` can't we just add these methods into `Subscription` ?

Comment (jai1) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/ServerCnx.java]: I felt that changing Interfaces is not good, they should be as concise as possible. Still if you want me to make add the methods to the interface let me know - i will make the changes.

Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/ServerCnx.java]: I think either we can cast to `PersistentSubscription` at line#272 or we can add this method to interface as these methods seem to be appropriate for the interface as well.

File: pulsar-common/src/main/proto/PulsarApi.proto [hunk 3]
```diff
@@ -285,9 +285,7 @@ message CommandPong {
 
 message CommandConsumerStats {
         required uint64 request_id         = 1;
-        required string topic_name         = 2;
-        required string subscription_name  = 3;
-        required uint64 consumer_id        = 4;
+        required uint64 consumer_id        = 2;
```
Comment (merlimat) [inline pulsar-common/src/main/proto/PulsarApi.proto]: You cannot change the tag id of existing fields and you cannot reuse the same id again.

Comment (merlimat) [inline pulsar-common/src/main/proto/PulsarApi.proto]: The tag id should still be 4 here

**Comment by jai1** [comment]
@merlimat - I agree that the tag id should remain 4 so that an old client can still talk to the new server but we don't have an old client. On the client side this feature will be released after this Pull request (i.e old clients don't have getBrokerConsumerStats API). I have added the version check to make sure that new client doesn't send requests to old broker.

Hence I don't think we need to keep the tag id as 4.

File: pulsar-common/src/main/proto/PulsarApi.proto (hunk 3 above)
Comment (rdhabalia) [inline pulsar-common/src/main/proto/PulsarApi.proto]: I think a new client can prevent requesting old broker by introducing version V8, but as we already had one release 1.17 with old -client which can send `topic_name=2` to new broker so, to handle this scenario we can keep `consumer_id=4` and comment-out removed fields for future reference.

**Comment by rdhabalia** [comment]
I think a new client can prevent requesting old broker by introducing version V8, but as we already had one release 1.17 with old -client which can send `topic_name=2` to new broker so, to handle this scenario we can keep `consumer_id=4` and comment-out removed fields for future reference.

**Comment by jai1** [comment]
We don't have any client released which uses the BrokerConsumerStats command 
a. Java - client PR is still not approved
b. CPP - PR was approved but not a single release has been made since 10th Jan 2017

**Comment by jai1** [comment]
Anyways I have  changed the field number to 4 in the last commit, I will merge the PR by tomorrow noon.

File: pulsar-common/src/main/proto/PulsarApi.proto (hunk 3 above)
Comment (jai1) [inline pulsar-common/src/main/proto/PulsarApi.proto]: Made the tag id 4

**Comment by jai1** [comment]
@merlimat - can you please give me a +1 on this - I have addressed your comments and got a +1 from Rajan but for some reason, the system is not allowing me to merge the changes - it is stuck on "merlimat requested changes"

**Comment by saandrews** [comment]
I think the consumer_id could have been 2 since there is no real client implementation released wth that change. Should we put it back to 2?

**Pull request merged**